text/3519-arbitrary-self-types-v2.md

```diff
@@ -0,0 +1,403 @@
+
+If you're implementing a smart pointer `P<T>` yet you can't allow a reference `&T` to exist, any option for implementing `Deref::deref` has drawbacks:
+
+* Specify `Deref::Target=T` and panic in `Deref::deref`. Not good.
+* Specify `Deref::Target=*const T`. This works with the current arbitrary self types feature, but is only possible if your smart pointer type contains a `*const T` which you can reference - this isn't the case for (for instance) weak pointers or types containing `NonNull`.
+
+Therefore, the current Arbitrary Self Types v2 provides a separate `Receiver` trait.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to be any type `T` where `T: Receiver<Target = Self>` or `Self`.
+
+The `Receiver` trait is simple and only requires to specify the `Target` type:
+
+```rust
+trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+The `Receiver` trait is already implemented for many standard library types:
+- smart pointers in the standard library: `Rc<Self>`, `Arc<Self>`, `Box<Self>`, and `Pin<Ptr<Self>>` (and in fact, any type which implements `Deref`)
+- references: `&Self` and `&mut Self`
+- pointers: `*const Self` and `*mut Self`
+
+Shorthand exists for references, so that `self` with no ascription is of type `Self`, `&self` is of type `&Self` and `&mut self` is of type `&mut Self`.
+
+All of the following self types are valid:
+
+```rust
+impl Foo {
+    fn by_value(self /* self: Self */);
+    fn by_ref(&self /* self: &Self */);
+    fn by_ref_mut(&mut self /* self: &mut Self */);
+    fn by_ptr(self: *const Self);
+    fn by_mut_ptr(self: *mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_custom_ptr(self: CustomPtr<Self>);
+}
+
+struct CustomPtr<T>(*const T);
+
+impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+}
+```
+
+## Recursive arbitrary receivers
+
+Receivers are recursive and therefore allowed to be nested. If type `T` implements `Receiver<Target=U>`, and type `U` implements `Receiver<Target=Self>`, `T` is a valid receiver (and so on outward).
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_rc_to_box(self: Rc<Box<Self>>) { ... }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `core` libs changes
+
+The `Receiver` trait is made public (removing its `#[doc(hidden)])` attribute), exposing it under `core::ops`. It gains a `Target` associated type.
+
+This trait marks types that can be used as receivers other than the `Self` type of an impl or trait definition.
+
+```rust
+pub trait Receiver {
+    type Target: ?Sized;
+}
+```
+
+A blanket implementation is provided for any type that implements `Deref`:
+
+```rust
+impl<P: ?Sized, T: ?Sized> Receiver for P
+where
+    P: Deref<Target = T>,
+{
+    type Target = T;
+}
+```
```

Comment:
There is actually a problem here that I didn't realize back when we talked about this. Implementing `Receiver` is effectively a promise to not add any new methods to the implementing type (given the reasons already outlined here, as adding methods could break downstream users for these kinds of types). This impl means, this promise now propagates to `Deref` implementations as well which seems like a very much unwanted side effect, especially given all the current `Deref` impls out there.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0ea4bedd4fc2cf5d3f0a633fbade68ac

Reply:
No because you can't add inherent impls to a foreign type, so you can't add new methods to them (ignoring fundamental types). With the `Receiver` trait you can kind of do this though.
```rs
//- crate: foo
pub struct Foo<T>(pub T);
impl<T> Receiver for Foo<T> {
    type Target = T;
}
//- crate: bar (depends on foo)
use foo::Foo;
struct Bar;
impl Bar {
    fn foobar(self: Foo<Bar>) {}
}
fn main() {
    Foo(Bar).foobar();
}
```
Adding a `fn foobar(self)` method to `Foo` would break the crate `bar` in this case, so adding a method to a `Receiver` implementing type is a breaking change. And since `Deref` here implies implementing `Receiver`, the same issue arises from just `Deref` implementations, that is the same issue would occur if instead of the `Receiver` impl in that example we had
```rs
impl<T> Deref for Foo<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target { &self.0 }
}
```

Reply:
Discarding `Receiver` from this RFC for the moment, I'm still missing what makes you take the stance that `Deref` implementors on today's Rust don't already carry the same promise not to add new methods. As far as I understand, this problem already exists with `Deref`, so adding `Receiver` to all `Deref` types changes nothing.

Reply:
Ah sorry I didn't really answer your question. Taking your example here and modifying it to be split across two crates (everything in one crate does not exhibit any semver problems for obvious reasons):
```rs

//- crate b
struct B;

impl B {
    fn foo(&self, x: i32) -> i32 {
        x
    }
}


//- crate: a depends on crate b
struct A(B);

impl std::ops::Deref for A {
    type Target = B;
    fn deref(&self) -> &B { 
        &self.0
    }
}

impl A {
    // As of Rust 1.73, this will shadow B::foo for A
    fn foo(&self) -> i32 {
        0
    }
}

fn main() {
    let a = A(B);
    a.foo(); // will always resolve to `A::foo` no matter the methods on `B` due to how method resolution works
}
```
Adding or removing the `foo` method has no effect here on the method resolution of objects of type `A`, as we will always discover the `A::foo` method first and then stop resolution as is. No crate that depends on these crates can later on "inject" new resolutions into this, unlike how the `Receiver` proposal here works

Reply:
Results from trying to reproduce this multi-crate method shadowing problem with the existing unstable `arbitrary_self_types` support:

```rust
// crate Foo
use core::ops::Deref;

pub struct Foo<T>(pub T);
impl<T> Deref for Foo<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
impl<T> Foo<T> {
    pub fn foobar(self) {
        println!("Calling foo's foobar");
    }
}


// crate bar
use foo::Foo;
struct Bar;
impl Bar {
    fn foobar(self: Foo<Bar>) {println!("Calling bar's foobar");}
}
fn main() {
    Foo(Bar).foobar();
}
```

Results:
```
error[E0034]: multiple applicable items in scope
 --> src/main.rs:8:14
  |
8 |     Foo(Bar).foobar();
  |              ^^^^^^ multiple `foobar` found
  |
note: candidate #1 is defined in an impl for the type `Bar`
 --> src/main.rs:5:5
  |
5 |     fn foobar(self: Foo<Bar>) {println!("Calling bar's foobar");}
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: candidate #2 is defined in an impl for the type `Foo<T>`
```

I think this does show that adding new methods to the `Foo` crate can break the `Bar` crate - even with the existing unstable `arbitrary_self_types`.

It seems to me that we can possibly solve this by amending the method resolution such that in ambiguous situations it always picks the one it would have picked prior to `arbitrary_self_types`, instead of reporting an ambiguity? (I'm not 100% following this thread though I must admit)

Reply:
> No crate that depends on these crates can later on "inject" new resolutions into this, unlike how the Receiver proposal here works

I'm not sure I'm following you fully, but I don't think there is a problem here. A user crate can indeed add something that looks like a new method on e.g. `Box<A>`, but because the crate that defines `Box` cannot know about `A`, there is nowhere where it could call a function that might be resolved as the downstream method.

Reply:
```rs
//- crate: foo
pub struct Foo<T>(pub T);
impl<T> Receiver for Foo<T> {
    type Target = T;
}
```

```rs
//- crate: bar (depends on foo)
use foo::Foo;
pub struct Bar;
impl Bar {
    pub fn foobar(self: Foo<Bar>) {}
}
```
```rs
//- crate: qux (depends on foo)
fn main() {
    Foo(Bar).foobar(); // resolves to `Bar::foobar`
}
```

Now, if crate `foo` adds the following:
```rs
impl<T> Foo<T> {
    pub fn foobar(self: Foo<T>) {}
}
```
crate `qux` will fail to compile due to an ambiguity, so adding new methods to a type that implements `Receiver` is always a breaking change. As `Foo(Bar).foobar()` could now be referring to either `Bar::foobar` or `<Foo::<Bar>>::foobar`

Since `Deref` implies implementing `Receiver` with the proposal here, that means the same restrictions fall onto any type implementing `Deref`. That's what I'm trying to get at here. That is in your `Box<A>` example, it doesn't need to to know about the `A`, because we can still reach this via generics.

Reply:
I guess the way out of this problem is to make `Receiver` opt-in, which means no blanket implementation for `Deref`. Then we need another mechanism to ensure that the targets of Receiver and Deref not diverge, but that probably can be achieved with ✨ compiler magic ✨.

Reply:
I think adding an inherent method `foo` to `T` where `T: Receiver<Target = U>` is a *minor* breaking change since anyone trying to call `U::foo` with `v.foo()` can just write `U::foo(v)` instead. (iirc this has precedence in deeming new trait methods as minor breaking changes even though they can become ambiguous and need `Trait::method(v)` instead of `v.method()`)

Therefore, I see no reason to remove the blanket `T: Receiver` implementation for `T: Deref`.

Reply:
I also agree that it's a breaking but non-major change. Adding a blanket impl for an existing trait, implementing a new method on an existing type, and newly implementing a trait for an existing type are all breaking changes (and ones deemed non-major) regardless of arbitrary self types.

Reply:
Oh I see. Doesn't `Deref` already have this exact problem though? That's exactly why `Box`/`Rc`/etc have no normal methods, since adding a `.foo()` would break any `Foo` that has `fn foo(&self)`

Reply:
I don't see how `Deref` already has this problem. `Box`, `Rc`, `Pin` etc have htis problem because they are already allowed as arbitrary receivers as they have been special cased. `Deref` has no say in that. With the RFC as written, all `Deref` implementations *will* have this problem. Anyways, even if this is considered a minor breakage only (which I'd argue against personally, it very much feels like a major breakage to me), the RFC should definitely talk about this, both in that implementing `Receiver` has this semver problem as well as that the blanket impl will extend it to `Deref` impls.

Reply:
No I mean plain Deref. Let me spell it out:

```rust
//- crate: foo (depends on nothing)
pub struct Foo<T>(pub T);
impl<T> Deref for Foo<T> {
    type Target = T;
    fn deref(...) {...}
}
```
```rust
//- crate: bar (depends on nothing)
pub struct Bar;
impl Bar {
    pub fn foobar(&self) {}
}
```
```rust
//- crate: qux (depends on bar and foo)
fn main() {
    Foo(Bar).foobar(); // resolves to `Bar::foobar`
}
```
Now, if crate foo adds the following:
```rust
impl<T> Foo<T> {
    pub fn foobar(&self) {}
}
```
then `Foo(Bar).foobar()` in crate qux now resolves to `Foo<T>::foobar`. I imagine this is a major breaking change since now qux could still compile yet be using a completely different function.

This is the same problem; `Deref` pointers already have to deal with it today. Hence I conclude: adding Receiver into the mix should not change anything, all is good.

Reply:
I agree with @Nadrieril, this is a non-issue for this RFC, since the problem is already present on `Deref` types.

The RFC actually [has a section that talks about this](https://github.com/adetaylor/rfcs/blob/arbitrary-self-types-v2/text/3519-arbitrary-self-types-v2.md#method-shadowing), I've updated that in https://github.com/rust-lang/rfcs/pull/3519#pullrequestreview-1733257247 to include a few more references to external resources that further explain the problem.

I'll also link to https://github.com/rust-lang/rust-clippy/issues/11820, which proposes a new clippy lint to catch inherent methods on smart pointers, to hopefully help implementers of such types (including in the future, implementers of `Receiver`) to avoid this pitfall.